examples_utils/notebook_logging/README.md

```diff
@@ -40,4 +40,42 @@ Unless logging is disabled, the following information will be collected:
 =============================================================================================================================================
 ```
 
-This is meant as a prototype to demonstrate how the capability could be implemented.
+## Design notes
+
+The following notes describe the design and architecture of the notebook logging module.
+
+### IPython extension format
+The module is written as an IPython extension for two reasons:
+- It can be loaded/unloaded into/out of the IPython kernel via IPython line magic easily and cleanly
+- It can access the IPython events register that we can register custom pre/post cell execution functions
+
+This allows us to perform event-based logging, where each cell execution counts as an event. Whilst some of the data we need to store is independent of the cells themselves, the majority of the information is specific and hence suits this logging method very well. 
+
+### Background processes
+Some functions that:
+- Have a very long runtime relative to the other pre/post cell execution methods
+- Only need to be run once per instance of notebook/logger (notebook metadata)
+- Need to access the notebook itself (JSON), which isnt saved on demand
+Need to be run in background processes which do not stall the execution of the pre/post cell execution functions to avoid making any delays visible to the user.
+
+Instead, the [multiprocess](https://docs.python.org/3/library/multiprocessing.html) library is used to create and manage the data structures and background processes. With this, we create, run, terminate and cleanup processes that are not exposed to the user and do not (significantly in any way) affect the python kernel itself.
+
```

Comment:
Is "multiprocess" a package, a class or something else? It sounds like it should be in code font to make this clearer.

Reply:
package, ok I'll put it inline in code